### ISSUE TITLE
Audit uses of `sym::cfg` and `sym::cfg_attr` (some ought to be replaced with `sym::cfg_trace` and `sym::cfg_attr_trace` resp.)

### ISSUE THREAD

**Issue by fmease** [rust-lang/rust-clippy#15982]
While working on cfg-related things in rustdoc, I noticed that clippy has 8 uses of `sym::cfg` and 2 uses of `sym::cfg_attr`. Most of these usages are probably fine but some should probably be replaced with `sym::cfg_trace` and `sym::cfg_attr_trace`, respectively.

Since https://github.com/rust-lang/rust/pull/138844 and https://github.com/rust-lang/rust/pull/138515 (March 2025), expanded active `#[cfg{,_attr}]`s get lowered to so-called *traces*. Meaning, when you're dealing with something that's running post-expansion, checking if the list of attributes of some given thing contains `sym::cfg` or `sym::cfg_attr` will always return `false`. For pre-expansion lints (etc.) on the other hand, that should still work.

For example I found fn `is_hir_ty_cfg_dependant`:

https://github.com/rust-lang/rust-clippy/blob/35e55a63f2b3da6a843af591618afeff21d98914/clippy_utils/src/lib.rs#L2275-L2282

which is to be used in `LateContext`s & HIR types (meaning: post-expansion) but it actually checks `sym::cfg` and `sym::cfg_attr`! This is quite clearly a bug! I wager that this function always returns `false`. Locally, I did modify it to check `cfg_trace` and `cfg_attr_trace` instead and no unit or UI tests failed, so this seems severely undertested.

As for concrete consequences / bugs, consider the following code which currently triggers the lint *char-lit-as-u8* despite `<Casts as LateLintPass>::check_expr` returning early if the cast-to type is cfg-dependent:

```rs
#![deny(clippy::char_lit_as_u8)]

#[cfg(true)]
type Char = u8;

fn main() {
    let _ = 'x' as Char; //~ char_lit_as_u8
}
```

With my local +`_trace` modification, the lint no longer fires which was the original intention of the lint pass author(s).

Indeed, if you replace `true` with `all()` and run `clippy-driver +nightly-2025-03-27 file.rs` (the last nightly before the aforementioned https://github.com/rust-lang/rust/pull/138844 landed) this compiles / the lint gets suppressed.

**Comment by fmease** [comment]
While it's quite easy to fix `is_hir_ty_cfg_dependant` and I could just do that myself right away, I don't have the time right now to investigate all other uses of this function and of `cfg{,_attr}` in general.

Moreover, it's a bit of a rabbit hole (which is too much for me rn). E.g., `unnecessary_cast::check` checks `is_ty_alias` before `is_hir_ty_cfg_dependant` (the latter always returns `false` as I've explained) for suppressing lints in cases like `as c_char` (`c_char` is cfg'ed to be either `i8` or `u8` depending on the system). Clearly, `is_hir_ty_cfg_dependant` isn't needed practically speaking, checking for a type alias seems to be more than sufficient esp. since you can't really define a cfg'ed ADT that you can also `as`-cast to. Even in the `char_lit_as_u8` example I gave above, you could simplify the lint that way.

Anyhow since that requires a bit of an investigation, I gave up for now.
